### DISCUSSION TITLE
Failed to highlight molecule differences with rdFMCS.FindMCS and rdRascalMCES.FindMCES

### DISCUSSION THREAD

**Discussion by cbwchuck** [rdkit/rdkit#7912] [Q&A]
Hi RDKit Community! 

I am trying to highlight differences of two molecules, I have tried `rdFMCS.FindMCS` and `rdRascalMCES.FindMCES` but found undesirable cases in both.

With FindMCS, cases like the following will kind of fail visually since identical ring will also be highlighted. But I assume algorithm-wise this is expected since FindMCS finds the maximum single substructure?

```python
from rdkit import Chem
from rdkit.Chem import Draw, AllChem, rdFMCS, MolFromSmiles, MolFromSmarts
from rdkit.Chem.Draw import IPythonConsole, rdDepictor
from rdkit.Chem.rdRascalMCES import FindMCES, RascalOptions
from rdkit.Chem.rdFMCS import FindMCS

def highlight_atoms(mol1, mol2, common_mol):
    # get atoms to highlight in mol1 and mol2
    match1 = mol1.GetSubstructMatch(common_mol)
    target_atm1 = []
    for atom in mol1.GetAtoms():
        if atom.GetIdx() not in match1:
            target_atm1.append(atom.GetIdx())

    match2 = mol2.GetSubstructMatch(common_mol)
    target_atm2 = []
    for atom in mol2.GetAtoms():
        if atom.GetIdx() not in match2:
            target_atm2.append(atom.GetIdx())

    return target_atm1, target_atm2

mol1 = Chem.MolFromSmiles("O=C(Nc1cn2ccnc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")
mol2 = Chem.MolFromSmiles("O=C(Nc1cc2c(c(N3CCC(F)(F)CC3)c1)OCC2)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")

mcs = FindMCS([mol1, mol2])
mcs_common_mol = MolFromSmarts(mcs.smartsString)

highlightAtomLists = highlight_atoms(mol1, mol2, mcs_common_mol)

Draw.MolsToGridImage([mol1, mol2], highlightAtomLists=highlightAtomLists, highlightBondLists=None)
```

<img width="406" alt="image" src="https://github.com/user-attachments/assets/591259c0-9936-4eb6-8798-6a47f2b8e1e0">

 While FindMCES solves the case above, cases like the following failed both methods and no difference was identified

```python
mol1 = Chem.MolFromSmiles("O=C(Nc1cn2ccnc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")
mol2 = Chem.MolFromSmiles("O=C(Nc1cn2cncc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")

mces = FindMCES(mol1, mol2)
mces_common_mol = MolFromSmarts(mces[0].smartsString)

highlightAtomLists = highlight_atoms(mol1, mol2, mces_common_mol)

Draw.MolsToGridImage([mol1, mol2], highlightAtomLists=highlightAtomLists, highlightBondLists=None)
```

<img width="399" alt="image" src="https://github.com/user-attachments/assets/4e7fa0fd-07f7-4eda-bd29-da68455eefcb">

Please kindly advice if this is due to my wrong usage of RDKit or maybe there is another method I should try.

Thanks in advance!

**Comment by DavidACosgrove** [comment]
Hi @cbwchuck,
This is an interesting one. I can only speak for the RASCAL code.   Below is the drawing of the two molecules with the bonds labelled:
![image](https://github.com/user-attachments/assets/9111ad91-110f-46ba-b65e-780fee225285)

It's important to note that the RASCAL algorithm finds the bonds in common to the 2 molecules, not the common atoms that FindMCS returns.  The results object from FindMCES returns the atoms and bonds that matched in the two molecules, so there's no need to do the SMARTS matching:
```
m1bonds = [b[0] for b in mces[0].bondMatches()]
m2bonds = [b[1] for b in mces[0].bondMatches()]
print(m1bonds)
print(m2bonds)
```
[0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45]
[0, 1, 2, 3, 4, 43, 8, 7, 6, 10, 44, 17, 16, 14, 15, 13, 12, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 42, 37, 36, 35, 34, 38, 39, 40, 41, 33, 5, 11, 45]
If you compare the two lists, you will see that bond 9 is missing and bond 5 in the first molecule has matched bond 43 in the second.

All the atoms in the molecule are involved in this set of bond matches, so there is nothing for your code to highlight as atom differences, so the picture produced is correct.

Having said all that, RASCAL has the option "completeAromaticRings" which defaults to True and is supposed to ensure that the matched bonds don't have partial aromatic rings in them, as is the case here.  I would have expected that the 2 6,5 heterocycles would have been excluded from the MCES.  I will take a look at it.  Could you raise an issue for this, for the RASCAL case only.

It's probable that the case for FindMCS is the same, but that's a different piece of code that will need to be considered by someone else.

Dave

**Comment by DavidACosgrove** [comment]
@cbwchuck 
I have spent some time looking at this, and I can't come up with a way of ruling out the solution produced without breaking a lot of other cases.  I will continue to think about it but at the moment I think I'll leave it as it is.
Dave

**Comment by cbwchuck** [comment]
Thanks so much for your explanation and effort Dave! @DavidACosgrove 

We explored this case a bit furtehr and thinking this might be an issue with isomer, and here are 2 more case may be helpful to inspire some thoughts. I will also raise an issue for our discussion here.

## Case 1
```python
mol1 = Chem.MolFromSmiles("O=C(Nc1cn2ccnc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")
mol2 = Chem.MolFromSmiles("O=C(Nc1cn2nccc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")
```

With MCES, it seems like a mis-highlighting at the atom between bond 6 and 7. Our understanding now is MCES matches all bonds but 5,6,7,8 and this is an expected result. However, visually it can be a bit misleading especially for non-chemical professionals (like us 😢 

<img width="403" alt="image" src="https://github.com/user-attachments/assets/00729e6e-ac54-4eb2-92bd-8c7ad0b5a8dd">

With MCS, it's even more misleading visually... after a closer look, our understanding is that the common structure on the left molecule was clockwise matching the cycle (bond 5,6,7,8,43) ,  and the right molecule was counter-clockwise matching the cycle.

<img width="396" alt="image" src="https://github.com/user-attachments/assets/00d3fa3d-d9d5-4bc5-afa2-c232ee1d58f9">

## Case 2
```python
mol1 = Chem.MolFromSmiles("O=C(Nc1cn2ccnc2c(N2CCC(F)(F)CC2)n1)c1ccc(NS(=O)(=O)CCO)cc1N1CCC2(CC1)CC2")
mol2 = Chem.MolFromSmiles("CS(=O)(=O)Nc1ccc(C(=O)Nc2cn3ccnc3c(N3CCC(F)(F)CC3)n2)c(N2CC3CCC(C3)C2)c1")
```

With MCES, it cannot detect the local isomer fragment on the top right

<img width="540" alt="image" src="https://github.com/user-attachments/assets/c76dd09f-ec92-44da-a867-d2ae576d9d76">


With MCS, it seems to be catching the difference, however, we cannot tell intuitively this should be the expected result...

<img width="540" alt="image" src="https://github.com/user-attachments/assets/c648f779-3c7f-41da-a900-463b46b4e446">
